### PR TITLE
[MOC] Implement LSTMCell fusion for Keras LSTM layers

### PR DESCRIPTION
**Pull request by rkazants**
**Details:** Implement LSTMCell fusion for Keras LSTM layers. Such sub-graph with split weights appears for bi-directional LSTM layer.
For custom model, we managed to fuse 6 LSTMCell sub-graphs and obtained 19% performance gain on CPU.

**Ticket:** 142924

### CODE REVIEW

**Review by evkotov — COMMENTED** [review]
Did you check a performance with and without this transformation? I found performance degradation? Should we use it only with LSTMSequenceFusion?

File: src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp [hunk 1]
```diff
@@ -352,3 +446,116 @@ ov::pass::LSTMCellFusion::LSTMCellFusion() {
...
+        // prepare weights in a format required by LSTMCell
+        // W weights must be in fico format
+        ov::Output<ov::Node> W = prepare_weight_fico(wf, wi, wc, wo, rg);
+        ov::Output<ov::Node> R = prepare_weight_fico(rf, ri, rc, ro, rg);
+
+        ov::Output<ov::Node> B = rg.make<ov::op::v0::Concat>(ov::OutputVector{bf, bi, bc, bo}, 0);
+        if (const auto& constant = ov::util::constantfold_subgraph(B))
+            B = constant;
+
+        auto lstm_cell = rg.make<op::v4::LSTMCell>(
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:541]: intel_cpu and intel_gpu plugins register their callbacks on LSTMFusion tranformations. These callbacks check if created LSTMCell is supported by backend. At least they check if LSTMCell weights are constants or could be the constants after folding. Please look at the CPU and GPU plugins to see these callbacks. 
Even if you add non-supported LSTMCell into the model there are several transformations (called from CPU/GPU plugins) that unroll this cell.
Consider please:
- add callback with new created LSTMCell
- check that after all transformations in CPU/GPU plugins we still have LSTMCell (not unrolled)

Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:541]: - add this transformation in CPU/GPU plugin to this list of transformation for appropriate callbacks

Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:541]: If these callbacks don't accept created LSTMCell, check that Loop optimization works and Constants are pushed into the loop. CPU/GPU plugin don't check outside loop

**Review by evkotov — CHANGES_REQUESTED** [review]
Did you check

File: src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp [hunk 2]
```diff
@@ -352,3 +446,116 @@ ov::pass::LSTMCellFusion::LSTMCellFusion() {
...
+        std::string h_activation_name = get_activation_name(ct_activated);
+
+        // prepare weights in a format required by LSTMCell
+        // W weights must be in fico format
+        ov::Output<ov::Node> W = prepare_weight_fico(wf, wi, wc, wo, rg);
+        ov::Output<ov::Node> R = prepare_weight_fico(rf, ri, rc, ro, rg);
+
+        ov::Output<ov::Node> B = rg.make<ov::op::v0::Concat>(ov::OutputVector{bf, bi, bc, bo}, 0);
+        if (const auto& constant = ov::util::constantfold_subgraph(B))
+            B = constant;
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:539]: There is a ConstantFolding transformation in MOC after this transformation. Do we need it here?

File: src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp [hunk 3]
```diff
@@ -23,6 +23,107 @@
...
+    // to get a shape equal to [hidden_size, input_size]
+    auto transpose_order = rg.make<ov::op::v0::Constant>(element::i32, ov::Shape{2}, std::vector<int32_t>{1, 0});
+    auto f_tr = rg.make<ov::op::v1::Transpose>(f, transpose_order);
+    auto i_tr = rg.make<ov::op::v1::Transpose>(i, transpose_order);
+    auto c_tr = rg.make<ov::op::v1::Transpose>(c, transpose_order);
+    auto o_tr = rg.make<ov::op::v1::Transpose>(o, transpose_order);
+
+    ov::Output<ov::Node> w = rg.make<ov::op::v0::Concat>(ov::OutputVector{f_tr, i_tr, c_tr, o_tr}, 0);
+    if (const auto& constant = ov::util::constantfold_subgraph(w)) {
+        w = constant;
```
Comment (evkotov) [inline src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:120]: There is a ConstantFolding transformation in MOC. Do we need it here?

**Review by evkotov — CHANGES_REQUESTED** [review]

File: src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp (hunk 1 above)
Comment (rkazants) [reply src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:624]: I combine these two transormations under one umbrella

File: src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp (hunk 2 above)
Comment (rkazants) [reply src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:622]: Yes, let us have it by the same way like in the other transformations.

File: src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp (hunk 3 above)
Comment (rkazants) [reply src/common/transformations/src/transformations/common_optimizations/lstm_cell_fusion.cpp:120]: Yes

**Comment by rkazants** [comment]
> Did you check a performance with and without this transformation? I found performance degradation? Should we use it only with LSTMSequenceFusion?

I got 19% performance gain.

**Review by jane-intel — APPROVED** [review]
Approved.

- It seems like the input_size dimension could be dynamic -- we could utilize symbol info in the future to make all the necessary checks

- General thought: LSTM fusion can be done in MOC, is it a reasonable next step? IR would be more compact. However, plugin callbacks won't be available there. So we would have to provide decomposition transformation for plugin-unsupported cases. This may be too much hassle for compact IR, so I'd be happy to discuss this further.

**Comment by rkazants** [comment]
> * General thought: LSTM fusion can be done in MOC, is it a reasonable next step? IR would be more compact. However, plugin callbacks won't be available there. So we would have to provide decomposition transformation for plugin-unsupported cases. This may be too much hassle for compact IR, so I'd be happy to discuss this further.

LSTMCell fusion is already triggered in MOC and IR gets compact. There is `LSTMCellDecomposition` transformations for plugins unsupporting `LSTMCell`.

> It seems like the input_size dimension could be dynamic -- we could utilize symbol info in the future to make all the necessary checks

For me, it makes sense and I can create a ticket for further improvement from symbolic optimization perspectives.

**Review by evkotov — APPROVED** [review]

**Pull request merged**